src/glob.zig

```diff
@@ -0,0 +1,2555 @@
+const PathLike = @import("./bun.js/node/types.zig").PathLike;
+const Maybe = @import("./bun.js/node/types.zig").Maybe;
+const Dirent = @import("./bun.js/node/types.zig").Dirent;
+const PathString = @import("./string_types.zig").PathString;
+const ZigString = @import("./bun.js/bindings/bindings.zig").ZigString;
+const isAllAscii = @import("./string_immutable.zig").isAllASCII;
+const EntryKind = @import("./bun.js/node/types.zig").Dirent.Kind;
+const Arena = std.heap.ArenaAllocator;
+const GlobAscii = @import("./glob_ascii.zig");
+const bun = @import("./bun.zig");
+const C = @import("./c.zig");
+
+const isWindows = @import("builtin").os.tag == .windows;
+
+const CodepointIterator = @import("./string_immutable.zig").PackedCodepointIterator;
+const Codepoint = CodepointIterator.Cursor.CodePointType;
+// const Codepoint = u32;
+const Cursor = CodepointIterator.Cursor;
+
+const CursorState = struct {
+    cursor: CodepointIterator.Cursor = .{},
+    /// The index in terms of codepoints
+    // cp_idx: usize,
+
+    fn init(iterator: *const CodepointIterator) CursorState {
+        var this_cursor: CodepointIterator.Cursor = .{};
+        _ = iterator.next(&this_cursor);
+        return .{
+            // .cp_idx = 0,
+            .cursor = this_cursor,
+        };
+    }
+
+    /// Return cursor pos of next codepoint without modifying the current.
+    ///
+    /// NOTE: If there is no next codepoint (cursor is at the last one), then
+    /// the returned cursor will have `c` as zero value and `i` will be >=
+    /// sourceBytes.len
+    fn peek(this: *const CursorState, iterator: *const CodepointIterator) CursorState {
+        var cpy = this.*;
+        // If outside of bounds
+        if (!iterator.next(&cpy.cursor)) {
+            // This will make `i >= sourceBytes.len`
+            cpy.cursor.i += cpy.cursor.width;
+            cpy.cursor.width = 1;
+            cpy.cursor.c = CodepointIterator.ZeroValue;
+        }
+        // cpy.cp_idx += 1;
+        return cpy;
+    }
+
+    fn bump(this: *CursorState, iterator: *const CodepointIterator) void {
+        if (!iterator.next(&this.cursor)) {
+            this.cursor.i += this.cursor.width;
+            this.cursor.width = 1;
+            this.cursor.c = CodepointIterator.ZeroValue;
+        }
+        // this.cp_idx += 1;
+    }
+
+    inline fn manualBumpAscii(this: *CursorState, i: u32, nextCp: Codepoint) void {
+        this.cursor.i += i;
+        this.cursor.c = nextCp;
+        this.cursor.width = 1;
+    }
+
+    inline fn manualPeekAscii(this: *CursorState, i: u32, nextCp: Codepoint) CursorState {
+        return .{
+            .cursor = CodepointIterator.Cursor{
+                .i = this.cursor.i + i,
+                .c = @truncate(nextCp),
+                .width = 1,
+            },
+        };
+    }
+};
+
+pub const GlobWalker = struct {
+    pub const Result = Maybe(void);
+
+    arena: Arena = undefined,
+
+    /// not owned by this struct
+    pattern: []const u8 = undefined,
```

Comment:
We have to be very careful whenever `undefined` is used 

Can this instead be an empty string?